### PR TITLE
Add a cloudscale_volume resource

### PR DESCRIPTION
**Pull request by davidhalter**
I have added a cloudscale_volume resource to make it possible to manage volumes in terraform.

I have added documentation, tests and I don't think there is more to do. Please let me know what you think. Thanks for looking at it!

Once this is merged, it would be really nice if we could release it :)

~ Dave

### CODE REVIEW

**Comment by ferdinandhuebner** [comment]
I would really like to see this merged. 

Is there anything I can help with to move this PR forward?

**Comment by davidhalter** [comment]
I have to wait for HashiCorp employees to take a look at it. If they don't react soon, I'll try their slack.

**Comment by davidhalter** [comment]
I tried to get them to review it, but no answer, yet. I will just write again after a while.

**Comment by davidhalter** [comment]
Since we're allowed to merge ourselves now, would you @ferdinandhuebner be able to review my code so we can merge it?

I think you probably know golang a bit better than my coworkers at cloudscale.ch, so I would prefer you reviewing my code!

**Comment by ferdinandhuebner** [comment]
Sure, I'll take a look at it next week.

**Review by ferdinandhuebner — COMMENTED** [review]
Before I take a closer look at the code, I have a question or observation about the way the volume resource is currently modelled and what the implications are, especially when it comes to deleting resources, servers in particular. 

At the moment, the volume resource is aligned with the cloudscale.ch API when it comes to attaching volumes to a server, i.e. the server where the volume is going to be attached to is specified on the volume resource. That makes sense.

For other providers, e.g. AWS, GCP, OpenStack or DigitalOcean, it's the other way around and you usually add the volume attachment to the server resource. 
With the OpenStack and DigitalOcean providers, there is a separate attachment resource, e.g. [compute_volume_attach_v2](https://www.terraform.io/docs/providers/openstack/r/compute_volume_attach_v2.html) that you could use to attach a volume to a server.

I'm not sure why, but I would have initially expected that the volume attachment happens on the server resource and the volume is just a thing that lies around waiting to be attached or destroyed.

I don't necessarily suggest that we use the same approach as the other providers, but I would like to ask what would happen if a server gets destroyed that has a volume attached to it. 

Does the volume get deleted like it is currently happening if I were to delete the server through the API? Do we want this behaviour in this case?

I think a separate volume attachment resource seems unnecessary at first, but it might help avoid accidental or unexpected deletion of a volume that is attached to a server. 

Consider this use-case: I create a server and a volume. The volume holds important data that I want to keep. I want to recreate the server for some reason. 
As a user, I would expect that I can just `terraform taint` the server and run `terraform apply`  to force re-creation. I don't know enough about resource dependencies in terraform, but would an attachment resource be the thing that would prevent the volume from being deleted by the cloudscale.ch API?

**Comment by davidhalter** [comment]
Good question. We know that this is still an issue with our API - I will discuss this internally. Thanks!

**Comment by ferdinandhuebner** [comment]
Thank you. Personally, I would think about making the current behaviour the concern of the GUI and not the API. On the GUI, you can ask the user for consent, e.g. "Do you want to delete volumes attached to this server?". It could then detach and delete the volumes and delete the server.

I thought about the attachment resource again and I don't see a real benefit for it. Whatever you decide on the API behaviour regarding server deletion, we could work around the current API behaviour on the terraform provider and detach all volumes prior to deleting a server.

I'll take a look at the code now.

File: website/docs/r/volume.html.markdown [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+# Create a new Server
+resource "cloudscale_server" "web-worker01" {
+  name        = "web-worker01"
+  flavor_slug = "flex-4"
+  image_slug  = "debian-9"
+  ssh_keys    = ["ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIL2jzgla23DfRVLQr3KT20QQYovqCCN3clHrjm2ZuQFW user@example.com"]
+}
+
+# Add a Volume to web-worker01
+resource "cloudscale_volume" "web-worker01-vip" {
```
Comment (ferdinandhuebner) [inline website/docs/r/volume.html.markdown]: Can you rename the resource to e.g. `web-worker01-volume` like in the `index.html.markdown`? The current name seems to be a copy/paste from the floating IP documentation.

File: website/docs/r/volume.html.markdown [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  server_uuids = ["${cloudscale_server.web-worker01.id}"]
+}
+```
+
+## Argument Reference
+
+The following arguments are supported when creating/changing Volumes:
+
+* `name` - (Required) Name of the new volume.
+* `size_gb` - (Required) The volume size in GB. Valid values are multiples of 50 for type "ssd" and multiples of 100 for type "bulk".
```
Comment (ferdinandhuebner) [inline website/docs/r/volume.html.markdown]: Update the valid increments for SSDs to 1 GB instead of 50 GB

File: website/docs/r/volume.html.markdown [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  type         = "ssd"
+  server_uuids = ["${cloudscale_server.web-worker01.id}"]
+}
+```
+
+## Argument Reference
+
+The following arguments are supported when creating/changing Volumes:
+
+* `name` - (Required) Name of the new volume.
```
Comment (ferdinandhuebner) [inline website/docs/r/volume.html.markdown:37]: Is there a restriction on the name? Can it be an arbitrary length? Is the name limited to a certain set of characters?

File: cloudscale/resource_cloudscale_volume_test.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "cloudscale_volume" {
+			continue
+		}
+
+		id := rs.Primary.ID
+
+		// Try to find the volume
+		_, err := client.Volumes.Get(context.Background(), id)
+		if err != nil && !strings.Contains(err.Error(), "Not found") {
```
Comment (ferdinandhuebner) [inline cloudscale/resource_cloudscale_volume_test.go]: Could you use the HTTP status code here instead of checking if the error message contains a string?

File: cloudscale/resource_cloudscale_volume_test.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+						"cloudscale_volume.basic", "server_uuids.#", "1"),
+					assertVolumeAttached(&server, &volume),
+				),
+			},
+			{
+				Config: serverConfig + "\n" + volumeConfig_detached(rInt2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudscaleVolumeExists("cloudscale_volume.basic", &volume),
+					resource.TestCheckResourceAttr(
+						"cloudscale_volume.basic", "server_uuids.#", "0"),
```
Comment (ferdinandhuebner) [inline cloudscale/resource_cloudscale_volume_test.go:122]: Not sure if it's necessary, but you could add an additional check against the cloudscale.ch API and check if the volume is indeed detached.

File: cloudscale/resource_cloudscale_volume_test.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+		*volume = *retrieveVolume
+
+		return nil
+	}
+}
+
+func assertVolumeAttached(server *cloudscale.Server, volume *cloudscale.Volume) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if server.UUID != (*volume.ServerUUIDs)[0] {
+			return fmt.Errorf("Server not properly attached to %s: %v", server.UUID, volume.ServerUUIDs)
```
Comment (ferdinandhuebner) [inline cloudscale/resource_cloudscale_volume_test.go]: Did you mean "volume not properly attached to %s: %v"? Would it make sense to rephrase it to something like "volume is not attached to %s but to %v"?

File: cloudscale/resource_cloudscale_volume.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+func fillVolumeResourceData(d *schema.ResourceData, volume *cloudscale.Volume) error {
+	d.Set("href", volume.HREF)
+	d.Set("name", volume.Name)
+	d.Set("size_gb", volume.SizeGB)
+	d.Set("type", volume.Type)
+
+	err := d.Set("server_uuids", volume.ServerUUIDs)
+	if err != nil {
+		log.Printf("[DEBUG] Error setting ssh_fingerprins attribute: %#v, error: %#v", volume.ServerUUIDs, err)
```
Comment (ferdinandhuebner) [inline cloudscale/resource_cloudscale_volume.go]: did you mean something like "error setting server_uuids attribute..."?

File: cloudscale/resource_cloudscale_volume.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+	}
+	return nil
+}
+
+func resourceVolumeRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudscale.Client)
+
+	volume, err := client.Volumes.Get(context.Background(), d.Id())
+	if err != nil {
+		if err.Error() == "detail: Not Found." {
```
Comment (ferdinandhuebner) [inline cloudscale/resource_cloudscale_volume.go]: Can you use the HTTP status code?

File: cloudscale/resource_cloudscale_volume.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+func resourceVolumeDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudscale.Client)
+	id := d.Id()
+
+	log.Printf("[INFO] Deleting Volume: %s", d.Id())
+	err := client.Volumes.Delete(context.Background(), id)
+
+	if err != nil && strings.Contains(err.Error(), "Not found") {
```
Comment (ferdinandhuebner) [inline cloudscale/resource_cloudscale_volume.go]: can you use the HTTP status code?

**Review by ferdinandhuebner — CHANGES_REQUESTED** [review]
I think the code looks goode, apart from some minor suggestions.

**Comment by davidhalter** [comment]
Hi Ferdinand

We finally decided to go with the behavior you initially suggested. Volumes will not be deleted in the API when a server is deleted.

Of course it might take us some time to implement it, because it might also have consequences for our GUI. Once that's done I will incorporate your feedback (good finds BTW) and we will merge this.

File: website/docs/r/volume.html.markdown (hunk 3 above)
Comment (davidhalter) [reply website/docs/r/volume.html.markdown:37]: It can be arbitrary. The length is limited, but not to something "small", so I would vote for leaving it like that. The API documentation is a bit more precise for quite a few things (we've chosen to do it like that when contributing `cloudscale_server`). It's valid criticism, though.

File: cloudscale/resource_cloudscale_volume_test.go (hunk 5 above)
Comment (davidhalter) [reply cloudscale/resource_cloudscale_volume_test.go:122]: I added additional checks against the cloudscale API in all places where volume existence is checked.

**Comment by davidhalter** [comment]
@ferdinandhuebner As we wrote in an email to you, we have changed our API to not delete volumes when servers are deleted. This is very beneficial for the terraform implementation and makes sense in general IMO. Thanks for bringing it up.

I fixed all the things you proposed (except one that I commented) and would like to merge if possible.

**Comment by ferdinandhuebner** [comment]
Looks good to me

**Review by ferdinandhuebner — APPROVED** [review]
